src/app/AttributeCache.cpp

```diff
@@ -219,5 +219,59 @@ CHIP_ERROR AttributeCache::GetStatus(const ConcreteAttributePath & path, StatusI
     return CHIP_NO_ERROR;
 }
 
+uint32_t AttributeCache::OnUpdateDataVersionFilterList(DataVersionFilterIBs::Builder & aDataVersionFilterIBsBuilder,
+                                                       DataVersionFilter * apDataVersionFilterList,
+                                                       size_t aDataVersionFilterListSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    TLV::TLVWriter backup;
+    uint32_t number = 0;
+
+    for (auto path1 = mChangedAttributeSet.begin(); path1 != mChangedAttributeSet.end(); path1++)
+    {
+        DataVersionFilter dataVersionFilter(path1->mEndpointId, path1->mClusterId, path1->mDataVersion);
+        aDataVersionFilterIBsBuilder.Checkpoint(backup);
+        VerifyOrExit(dataVersionFilter.IsValidDataVersionFilter(), err = CHIP_ERROR_INVALID_ARGUMENT);
+        for (size_t index = 0; index < aDataVersionFilterListSize; index++)
+        {
+            if (apDataVersionFilterList[index].IsSameCluster(*path1))
+            {
+                continue;
+            }
+        }
```

Comment:
I don't understand this loop.  It's a no-op, no?

Reply:
if user provides their data version list, which is intersecting with cached one, then we prefer used-defined one, and skip the cached one here.

Reply:
Was this code tested?  Because what this code does not not match the description just given here.

Reply:
yes, it is tested.

Reply:
TestReadWithAttributeCacheWithUserDefinedDataVersionFilter, the test provides one user-defined data version filter(version is different from the one in handler), then I issue two related read which has two path with different attribute id, and their endpoint/cluster id is same as user-defined data version filter, the first read get two attribute data IB, then attributeCache caches these two paths along with their versions, second read uses the above code, and find cached data version has same cluster as the user-defined data version filter has, and we skip the cached data  version, finally this read get two attributeData IBs.

Reply:
fixed, thanks